### PR TITLE
Opportunity Detection

### PR DESCRIPTION
**Pull request by ruidazeng**
Opportunity Detection algorithms implemented

### CODE REVIEW

**Review by samdevo — APPROVED** [review]
Nice!

**Comment by samdevo** [comment]
@ruidazeng so is this just https://github.com/Drakkar-Software/Triangular-Arbitrage/blob/master/triangular_arbitrage/detector.py but with some modifications to allow larger cycles ? That's okay I guess, but I believe they did it for a reason. Removing the triangular requirement will just give you a super bad runtime since the set of *all* cycles is enormous

**Comment by ruidazeng** [comment]
> @ruidazeng so is this just https://github.com/Drakkar-Software/Triangular-Arbitrage/blob/master/triangular_arbitrage/detector.py but with some modifications to allow larger cycles ? That's okay I guess, but I believe they did it for a reason. Removing the triangular requirement will just give you a super bad runtime since the set of _all_ cycles is enormous

I don't think the set of all cycle is for a specific set is significantly larger than the set of all triangular cycles for exchanges that only contains more niche currencies. A big reason why they only cared about triangular arbitrages is that for large centralized exchanges (like Binance, Coinbase, KuCoin, etc) is that if there is an exchange ticker available between every currency to each other, then most longer cycle can be composed entirely from the smaller 3-cycles (triangular cycles), hence higher order cycle detections are not necessary.

However, for smaller DEX that concerns with super niche currencies, where links do not necessarily exist between every currencies (like the example below, very simplified of course and completely made up), then the triangular will completely miss the arbitrage opportunity, even when if it's extremely profitable:

dex

Either way, the algorithm is the same - [Johnson's Algorithm (1975)](https://dl.acm.org/doi/pdf/10.1145/321992.321993) for finding the shortest paths between all pairs of vertices in an edge-weighted directed graph.

File: src/opportunity_detector.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+            reversed_ticker = ShortTicker(Symbol(ticker.symbol.quote, ticker.symbol.base),
+                                          1 / ticker.last_price, reversed=True)
+            graph.add_edge(ticker.symbol.quote, ticker.symbol.base, ticker=reversed_ticker)
+
+    best_profit = 0
+    best_triplet = None
+
+    # Find all cycles in the graph (specifically triplets)
+    for cycle in nx.simple_cycles(graph):
+        if len(cycle) != 3:
```
Comment (ruidazeng) [inline src/opportunity_detector.py:54]: @samdevo If you look at the original code from their triangular arbitrage opportunity detection, the time complexity is the same as the algorithm is the same.

**Pull request merged**